restord/ImageSessionServer.m

```diff
@@ -366,24 +372,69 @@ - (NSURL *)mountDisk:(DADiskRef)disk {
                            withIntermediateDirectories:YES
                                             attributes:nil
                                                  error:NULL];
-  DADiskMount(disk, (__bridge CFURLRef)directoryURL, 0,
-              &MountUnmountCallback, (__bridge void *)sema);
+  DADiskMount(disk, (__bridge CFURLRef)directoryURL, 0, &MountUnmountEjectCallback,
+              (__bridge void *)sema);
   if (dispatch_semaphore_wait(sema, dispatch_time(DISPATCH_TIME_NOW, 5 * 60 * NSEC_PER_SEC))) {
     NSLog(@"%@ Timed out while mounting disk", self);
     return nil;
   }
-
+  [self disableSpotlight:directoryURL];
   NSDictionary *desc = CFBridgingRelease(DADiskCopyDescription(disk));
   return desc[(__bridge NSString *)kDADiskDescriptionVolumePathKey];
 }
 
-// Unmount a given disk.
-- (void)unmountDisk:(DADiskRef)disk {
+// Unmount a given disk/whole disk.
+- (void)unmountDisk:(DADiskRef)disk withOptions:(DADiskUnmountOptions)options {
+  dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+  if (options & kDADiskUnmountOptionWhole) {
+    disk = DADiskCopyWholeDisk(disk);
+  }
+  DADiskUnmount(disk, options, &MountUnmountEjectCallback, (__bridge void *)sema);
+  if (dispatch_semaphore_wait(sema, dispatch_time(DISPATCH_TIME_NOW, 5 * 60 * NSEC_PER_SEC))) {
+    NSLog(@"%@ Timed out while unmounting disk: %@", self, disk);
+  }
+  if (options & kDADiskUnmountOptionWhole) {
+    CFRelease(disk);
+  }
+}
+
+// Eject the whole disk for a given diskref.
+- (void)ejectDisk:(DADiskRef)disk {
   dispatch_semaphore_t sema = dispatch_semaphore_create(0);
-  DADiskUnmount(disk, kDADiskUnmountOptionDefault, &MountUnmountCallback, (__bridge void *)sema);
+  DADiskRef wholeDisk = DADiskCopyWholeDisk(disk);
+  DADiskEject(wholeDisk, kDADiskEjectOptionDefault, &MountUnmountEjectCallback,
+              (__bridge void *)sema);
   if (dispatch_semaphore_wait(sema, dispatch_time(DISPATCH_TIME_NOW, 5 * 60 * NSEC_PER_SEC))) {
     NSLog(@"%@ Timed out while unmounting disk", self);
   }
+  CFRelease(wholeDisk);
+}
+
+// spotlight seems to start up very quickly once the disk is mounted.  Tell it to stop.
+- (int)disableSpotlight:(NSURL *)mountPoint {
+  NSLog(@"Running /usr/bin/mdutil -i off %@", mountPoint.path);
+  NSTask *mdUtil = [[NSTask alloc] init];
+  mdUtil.standardOutput = [NSPipe pipe];
+  mdUtil.launchPath = @"/usr/bin/mdutil";
+  mdUtil.arguments = @[ @"-i", @"off", @"-d", mountPoint.path ];
+  [mdUtil launch];
+  [mdUtil waitUntilExit];
+  NSData *sout = [[mdUtil.standardOutput fileHandleForReading] readDataToEndOfFile];
```

Comment:
nit: `NSData *sout = [mdUtil.standardOutput.fileHandleForReading readDataToEndOfFile]`

Reply:
done, changed throughout.

Reply:
Actually, because standardOutput is an id, and can be either an NSPipe or an NSFileHandle, can't use this as a property.

Reply:
Ah, right!